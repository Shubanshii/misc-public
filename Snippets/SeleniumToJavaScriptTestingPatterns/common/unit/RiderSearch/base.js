import React from 'react'
import PageObjectBasePage from '../../../_patterns/page-object/poBasePage.pattern'

import pageElemMap from './elementmap'
import RiderSearchComponent from '../../../../client/components/rider-search/RiderSearch'


import { riderSearchProps } from '../../../_mocks/common/RiderSearch'

export class RiderSearch extends PageObjectBasePage
{
  constructor(validatorType,renderMethod='mount')
  {
    super({
      component:<RiderSearchComponent {...riderSearchProps} />,
      pageElemMap,
      pageValidator:validatorType,
    },renderMethod);
  }
}
