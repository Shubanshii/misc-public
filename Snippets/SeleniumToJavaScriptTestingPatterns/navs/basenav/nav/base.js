import React from 'react'
import { Provider } from 'react-redux'
import configureStore from '../../../../client/store/configureStore.dev'

import propsSingleton from '../../../_patterns/singleton/singletonBase.pattern'

import pageElemMap from './elementmap'
import PageObjectBasePage from '../../../_patterns/page-object/poBasePage.pattern'
import {sinonNavActions, sinonNavProps} from './sinonBase'

import Footer from '../../../../client/components/common/Footer'
import Header from '../../../../client/components/common/Header'
import RiderSearch from '../../../../client/components/rider-search/RiderSearch'
import VerticalNav from '../../../../client/components/common/VerticalNav'

import AlertCards from '../../../../client/components/alerts/AlertCards'


import { topNavProps,sidebarProps,footerProps,alertCardsProps } from '../../../_mocks/navs/basenav'

export class TopNav extends PageObjectBasePage
{
  constructor(validatorType,renderMethod='mount')
  {
    super({
      component:<Header {...topNavProps} changeRoute={(param)=>{console.log('changing route',param);}} />,
      pageElemMap,
      pageValidator:validatorType,
    },renderMethod);
  }
}

export class SideBar extends PageObjectBasePage
{
  constructor(validatorType)
  {
    super({
      component:<VerticalNav {...sidebarProps}/>,
      pageElemMap,
      pageValidator:validatorType
    },'render');
  }
}
export class SideBarPII extends PageObjectBasePage
{
  constructor(validatorType)
  {
    super({
      component:<VerticalNav {...sidebarProps} user={{...sidebarProps.user,user_type:'PII'}} />,
      pageElemMap,
      pageValidator:validatorType
    },'render');
  }
}

export class FooterNav extends PageObjectBasePage
{
  constructor(validatorType)
  {
    super({
      component:<Footer {...footerProps} />,
      pageElemMap,
      pageValidator:validatorType
    },'render');
  }
}

export class FooterNavNoReload extends PageObjectBasePage
{
  constructor(validatorType)
  {
    super({
      component:<Footer {...footerProps} user={{...footerProps.user,email:'no_reload'}} />,
      pageElemMap,
      pageValidator:validatorType
    },'render');
  }
}

export class Alerts extends PageObjectBasePage
{
  constructor(validatorType,renderMethod='mount')
  {
    const alertsProps = propsSingleton.get('fullPageData')

    alertsProps.actions = { ...sinonNavActions } // <-- to add to nested props

    super({
      component:<AlertCards {...alertsProps} {...sinonNavProps} />,
      pageElemMap,
      pageValidator:validatorType
    },renderMethod);
  }
}
