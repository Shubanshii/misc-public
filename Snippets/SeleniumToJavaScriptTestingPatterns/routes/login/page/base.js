import React from 'react'
import { Provider } from 'react-redux'
import configureStore from '../../../../client/store/configureStore.dev';

import PageObjectBasePage from '../../../_patterns/page-object/poBasePage.pattern'

import pageElemMap from './elementmap'
import pageValidator from './validator'

import AuthForm from '../../../../client/components/user-auth/AuthForm';
import { authProps } from '../../../_mocks/pages/authroute/props';

import { inputChangeHelper } from '../../../_utils/enzyme.utils';

export default class LoginBase extends PageObjectBasePage
{
  constructor()
  {
    super({
      component:<AuthForm {...authProps} />,
      pageElemMap,
      pageValidator
    });
  }

  formFocus()
  { this.Map.inputUser.simulate('focus'); }
  changeInputVal(o)
  {
    this.Map.inputUser.simulate('change', inputChangeHelper(o.username));
    this.Map.inputPassword.simulate('change',  inputChangeHelper(o.pass) );
  }
  submitForm()
  { this.Map.loginSubmitButton.simulate('click') }
}
