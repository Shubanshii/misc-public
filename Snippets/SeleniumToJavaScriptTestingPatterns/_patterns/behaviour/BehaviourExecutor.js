export default class BehaviourExecutor
{
  Execute(behaviours=[])
  {
    let exec;
    function* generateExecution()
    {
      for(let i=0; i<behaviours.length; i++)
      { if(behaviours[i]){ yield behaviours[i].Execute() } }
    }
    function execNext()
    {
      const yielded = exec.next();
      if(!yielded.done)
      { execNext(); }
    }
    exec = generateExecution();
    execNext();
  }
}
