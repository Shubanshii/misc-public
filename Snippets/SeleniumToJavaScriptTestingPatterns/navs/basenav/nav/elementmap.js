import PageObjectBaseMap from '../../../_patterns/page-object/poBaseElementMap.pattern'

import AlertCards from '../../../../client/components/alerts/AlertCards'

export default class PageElementMap extends PageObjectBaseMap
{
  //Getters
  get current()
  { return this.props }

    //Header
  get header()
  { return this.props.find('header') }
  get alerts()
  { return this.props.find('header .alerts-block') }
  get alertCard()
  { return this.props.find('header #alertCards') }
  get alertCardComponent()
  { return this.props.find(AlertCards) }
  get alertsAmt()
  { return this.props.find('header .alerts-block .alerts-num') }
  get logo()
  { return this.props.find('header a.logo img').first(); }
  get input()
  { return this.props.find('header form.header-search input#rider-search-home').first(); }
  get inputFromSearchComponent()
  { return this.props.find('input#rider-search-home').first(); }
  get notifications()
  { return this.props.find('header .alerts-block').first(); }
  get notificationAmt()
  { return this.props.find('header .alerts-block .alerts-num').first(); }
  get userSection()
  { return this.props.find('header .user-block').first(); }
  get userLogout()
  { return this.props.find('header .user-block ~ .dropdown-menu .menu-item').first(); }

    //Sidebar
  get mobileCloseMenu()
  { return this.props.find('.vertical-nav .close-menu .fa-times').first(); }
  get mobileUserInfo()
  { return this.props.find('.vertical-nav .user-info').first(); }
  get mobileUserLogo()
  { return this.props.find('.vertical-nav .user-info .user-logo').first(); }
  get mobileLogOut()
  { return this.props.find('.vertical-nav svg#icon-power-logout').first(); }
  get mobileFooterLinks()
  { return this.props.find('.vertical-nav footer#footer .link-wrapper').first(); }
  get mobileCopyRight()
  { return this.props.find('.vertical-nav footer#footer .cr p').first(); }

  get allRides()
  { return this.props.find('.vertical-nav svg#icon-calendar').first(); }
  get analytics()
  { return this.props.find('.vertical-nav svg#icon-chart-areaspline').first(); }
  get settings()
  { return this.props.find('.vertical-nav svg#icon-settings').first(); }
  get invoice()
  { return this.props.find('.vertical-nav svg#icon-currency-usd').first(); }
  get iconsWrapper()
  { return this.props.find('.vertical-nav .icons-wrapper').first();}

    //Footer
  get footerCopyRight()
  { return this.props.find('.app-container footer#footer .cr p').first(); }
  get footerFooterLinks()
  { return this.props.find('footer#footer .link-wrapper').first(); }

  get logoLink()
  { return this.props.find('header a.logo').first(); }

  //From Alerts component
  get directAlerts()
  { return this.props.find('.alerts-block'); }
  get directAlertCard()
  { return this.props.find('#alertCards'); }
  get directAlertCardViewAll()
  { return this.props.find('.heading > a'); }
  get directAlertCardTitle()
  { return this.props.find('#alertCards h3'); }
  get directAlertCardRider()
  { return this.props.find('#alertCards p.rider-name'); }
  get directAlertNextButton()
  { return this.props.find('#alertCards .header-alert-right-arrow'); }
  get directAlertPrevButton()
  { return this.props.find('#alertCards .fa-chevron-left').parent(); }
}
