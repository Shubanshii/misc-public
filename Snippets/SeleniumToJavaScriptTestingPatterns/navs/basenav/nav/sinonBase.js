import SinonSingleton from '../../../_patterns/singleton/singletonBase.pattern'
import { sinonUtils } from '../../../_utils/sinon.utils'
import { alertResponseObjMock } from '../../../_mocks/navs/sinonmocks'

  //The object below should contain all props that will be spied/stubbed/mocked for a specific component
const SinonBaseListeners = {
  toggleAlertsCards: sinonUtils.spy.set,
  updateRoute: sinonUtils.spy.set,
  getAlerts:  sinonUtils.stub.set
}

//instantiate listeners
const toggleAlertsCardsSpy = SinonBaseListeners.toggleAlertsCards()
const updateRouteSpy = SinonBaseListeners.updateRoute()
const getAlertsStub = SinonBaseListeners.getAlerts()

//set stub/mock data
getAlertsStub.onCall(0).callsArgWith(1,alertResponseObjMock.alertFailedType)
             .onCall(2).callsArgWith(1,alertResponseObjMock.alertWarningLevelOne);

//add to singleton
sinonUtils.save(toggleAlertsCardsSpy,'alertsSinon')
sinonUtils.save(updateRouteSpy,'alertsUpdateRouteSinon')
sinonUtils.save(getAlertsStub,'getAlertsSinon')

//organize by common action methods
const navActionSpiesObj = {
  updateRoute:updateRouteSpy
}
const navActionStubsObj = {
  getAlerts:getAlertsStub
}
const navActionMocksObj = { }


//ogranize by methods called from props
const navPropSpiesObj = {
  toggleAlertsCards:toggleAlertsCardsSpy
}
const navPropStubsObj = { }
const navPropMocksObj = { }


export const sinonNavActions = Object.assign({}, navActionSpiesObj, navActionStubsObj, navActionMocksObj);
export const sinonNavProps = Object.assign({}, navPropSpiesObj, navPropStubsObj, navPropMocksObj);
