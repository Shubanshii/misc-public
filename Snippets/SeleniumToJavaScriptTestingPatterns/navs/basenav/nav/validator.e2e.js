import PageObjectValidator from '../../../_patterns/page-object/poBaseValidator.pattern.js';
import SinonObj from '../../../_patterns/singleton/singletonBase.pattern'

import navPageElemMap from './elementmap';

export default class NavValidatorE2E extends PageObjectValidator
{

  constructor(mountedObj)
  {
    super(new navPageElemMap(mountedObj))
  };

  dispatchActionOnClick(action)
  {
    for(let i=0; i<action.correctActionReturns.length; i++ )
    {
      expect(action.correctActionReturns[i]).toEqual(SinonObj.get(i).currentSpyValue[0])
    }
  }

}
