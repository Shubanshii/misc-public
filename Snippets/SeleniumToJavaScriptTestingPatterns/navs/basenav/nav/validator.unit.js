import PageObjectValidator from '../../../_patterns/page-object/poBaseValidator.pattern.js';

import navPageElemMap from './elementmap';

export default class NavValidatorUnit extends PageObjectValidator
{
  constructor(mountedObj)
  { super(new navPageElemMap(mountedObj)) };

  hasLogo()
  { expect(this.Map.logo).toBeDefined(); }
  hasInputField()
  { expect(this.Map.input).toBeDefined(); }
  hasNotificationIcon()
  { expect(this.Map.notifications).toBeDefined(); }
  hasUserSection()
  { expect(this.Map.userSection).toBeDefined(); }

  hasMobileCloseMenu()
  { expect(this.Map.mobileCloseMenu).toBeDefined(); }
  hasMobileUserLogo()
  { expect(this.Map.mobileUserLogo).toBeDefined(); }
  hasMobileLogOut()
  { expect(this.Map.mobileLogOut).toBeDefined(); }
  hasMobileFooterLinks()
  { expect(this.Map.mobileFooterLinks).toBeDefined(); }
  hasMobileCopyRight()
  { expect(this.Map.mobileCopyRight).toBeDefined(); }

  hasAllRides()
  { expect(this.Map.allRides).toBeDefined(); }
  hasAnalytics()
  { expect(this.Map.analytics).toBeDefined(); }
  hasSettings()
  { expect(this.Map.settings).toBeDefined(); }
  hasInvoice()
  { expect(this.Map.invoice).toBeDefined(); }

  hasFooterCopyRight()
  { expect(this.Map.footerCopyRight).toBeDefined(); }
  hasFooterFooterLinks()
  { expect(this.Map.footerFooterLinks).toBeDefined(); }


  hasCorrectElements(curNav)
  {
    if(curNav === 'topNav')
    {
      this.hasLogo();
      this.hasInputField();
      this.hasNotificationIcon();
      this.hasUserSection();
    }
    if(curNav === 'sideBar')
    {
      this.hasMobileUserLogo();
      this.hasMobileLogOut();
      this.hasMobileFooterLinks();
      this.hasMobileCopyRight();
      this.hasAllRides();
      this.hasAnalytics();
      this.hasSettings();
      this.hasInvoice();

      this.showsCalendarLink(false);
    }
    if(curNav === 'sideBarPII')
    { this.showsCalendarLink(false); }
    if(curNav === 'footer')
    {
      this.hasFooterCopyRight();
      this.hasFooterFooterLinks();

      this.showsReloadDatabase(true);
    }
    if(curNav === 'footerNoReload')
    { this.showsReloadDatabase(false); }
  }
  showsCalendarLink(b)
  {
    (b)
      ? expect(~this.Map.iconsWrapper.html().indexOf('icon-calendar')).not.toEqual(0)
      : expect(~this.Map.iconsWrapper.html().indexOf('icon-calendar')).toEqual(0)
  }
  showsReloadDatabase(b)
  {
    (b)
      ? expect(~this.Map.footerFooterLinks.html().indexOf('Reload')).not.toEqual(0)
      : expect(~this.Map.footerFooterLinks.html().indexOf('Reload')).toEqual(0)
  }
  hasCorrectProps(curNav)
  {
    if(curNav === 'topNav')
    {
      let prps = this.Map.input.get(0);
      expect(prps.name).toEqual('input');
      expect(prps.attribs.name).toEqual('riderSearch');
      expect(prps.attribs.type).toEqual('text');
      expect(prps.attribs.label).toEqual('Search by Rider Name, DOB, Rider or Ride ID#, Phone or Address');
    }
    if(curNav === 'sideBar')
    {
      const userName = this.Map.mobileUserInfo.find('p').get(0);
      expect(userName.children[0].data).toEqual('Mary Irving');
    }
  }
  alertVisible(b)
  {
    (b)
      ? expect(this.Map.alertsAmt.text()).not.toEqual('')
      : expect(this.Map.alertsAmt.text()).toEqual('')
  }
  accountHasUserInfo()
  {
    console.log('Example using find instead of extending the Map Class');
    expect(this.Map.userSection.find('.user-icon').text()).toEqual('M')
    expect(this.Map.userSection.find('.user-name').text().replace(/\W/,'')).toEqual('MaryIrving')
  }

}
