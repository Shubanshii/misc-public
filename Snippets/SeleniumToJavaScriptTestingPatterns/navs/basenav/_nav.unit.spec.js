import * as rTD from '../../_utils/realTimeData.utils'

import BehaviourExecutor from '../../_patterns/behaviour/BehaviourExecutor';
import unitValidator from './nav/validator.unit'

import NavigationSingleton from '../../_patterns/singleton/singletonBase.pattern.js'

//Behaviours
  //We will probably never have to use behaviours with unit tests.
    //But on the off chance we need to, it is possible using the below as an example
import AssertHasAllElements from './nav/behaviours/assert.hasAllElements.behaviour'
import AssertHasAllProps from './nav/behaviours/assert.hasAllProps.behaviour'
import AlertIsVisible from './nav/behaviours/assert.AlertIsVisible.behaviour'

//helpers
import CommonActionPopulateInputTextField from '../../_common/behaviours/action.populateInputTextField.behaviour'

import { TopNav,SideBar,SideBarPII,FooterNav,FooterNavNoReload } from './nav/base'

beforeAll((done) => {
  rTD.setUserCredentials()
    .then(()=>{ done(); })
    .catch((err) => { console.error(err); done();} )
});

describe('Site Navigation', () => {
  let behaviourExecutor,
      navBase;

  beforeEach((done) => {
    behaviourExecutor = new BehaviourExecutor()

    rTD.setPageProperties('nav')
      .then(()=>
      {
        NavigationSingleton.remove('fullPageData')
        const fullPageData =rTD.mergeState(NavigationSingleton.get('userData').app, NavigationSingleton.get('pageData').app)
        NavigationSingleton.add(fullPageData,'fullPageData')
        done();
      })
      .catch((err) => { console.error(err); done();} )
  })

  describe('Top Bar Navigation', () => {
    beforeEach(() => navBase = new TopNav(unitValidator,'render'));

    test.skip('Correct Base Objects Exists', () =>{
      behaviourExecutor.Execute([
        new AssertHasAllElements(navBase,'topNav'),
        new AssertHasAllProps(navBase,'topNav'),
      ])
    })

    test.skip('Alert Should Display Number of Alerts if > 0', () =>{
      behaviourExecutor.Execute([ new AlertIsVisible(navBase,true) ]);
    })
    test('Alert Number and Circle Should be Hidden if Zero Alerts', () =>{
      behaviourExecutor.Execute([ new AlertIsVisible(navBase,false) ]);
    })

    test('Account Link Should Display Avatar & User Name', ()=>{
      navBase.Validate().accountHasUserInfo();
    })

    test('Layout is unchanged', () =>
    {
      //This may not be the best example depending on how we handle the live data propogation
      //(e.g. login name may change - but demonstrates usage until then)
      navBase.Snap()
    })

  })

  describe('Side Bar Navigation', () => {
    beforeEach(() => navBase = new SideBar(unitValidator));

    test('Correct Elements Exist', () =>{
      navBase.Validate().hasCorrectElements('sideBar')//validate is just a method - can accept args
      navBase.Validate().hasCorrectProps('sideBar')
    })
    test('Icons Based on User Type ', () =>{
      navBase = new SideBarPII(unitValidator)//overwrite properties to change user_type
      navBase.Validate().hasCorrectElements('sideBarPII')
    })
  })

  describe('Footer Navigation', () => {
    beforeEach(() => navBase = new FooterNav(unitValidator));

    test('Correct Elements Exist', () =>{
      navBase.Validate().hasCorrectElements('footer')
    })
    test('Reload Database Should Not Be Visible Outside of Dev',()=>{
      navBase = new FooterNavNoReload(unitValidator)//overwrite properties to change user_type
      navBase.Validate().hasCorrectElements('footerNoReload')
    })
  })
});
