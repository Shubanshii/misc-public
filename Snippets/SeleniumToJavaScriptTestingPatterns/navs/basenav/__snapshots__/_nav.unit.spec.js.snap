// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site Navigation Top Bar Navigation Layout is unchanged 1`] = `
<header
  className="relative-position"
  onKeyDown={[Function]}
>
  <i
    aria-hidden="true"
    className="fa fa-2x fa-bars"
    onClick={[Function]}
  />
  <div
    className="left-side-block"
  >
    <a
      className="logo"
      onClick={[Function]}
    >
      <img
        alt="Circulation logo"
        src="/dist/images/logo-c.png"
      />
    </a>
  </div>
  <div
    className="right-side-block"
  >
    <div
      className="user-block "
      onClick={[Function]}
    >
      <div
        className="user-icon"
      >
        M
      </div>
      <span
        className="user-name"
      >
        Mary
        Â 
        Irving
      </span>
      <div
        className="triangle"
      />
    </div>
  </div>
  <div
    className="mobile-search input "
  />
</header>
`;
